application/models/services/SurveyUpdater/GeneralSettings.php

```diff
@@ -0,0 +1,570 @@
+<?php
+
+namespace LimeSurvey\Models\Services\SurveyUpdater;
+
+use Survey;
+use Permission;
+use LSYii_Application;
+use PluginEvent;
+use Date_Time_Converter;
+use LimeSurvey\SessionData;
+use LimeSurvey\PluginManager\PluginManager;
+use LimeSurvey\Models\Services\Exception\{
+    ExceptionPersistError,
+    ExceptionNotFound,
+    ExceptionPermissionDenied
+};
+
+/**
+ * Service GeneralSettings
+ *
+ * Service class for survey language setting updating.
+ *
+ * Dependencies are injected to enable mocking.
+ * @SuppressWarnings(PHPMD.ExcessiveClassComplexity)
+ */
+class GeneralSettings
+{
+    private Permission $modelPermission;
+    private Survey $modelSurvey;
+    private LSYii_Application $yiiApp;
+    private SessionData $sessionData;
+    private PluginManager $yiiPluginManager;
+    private LanguageConsistency $languageConsistency;
+
+    const FIELD_TYPE_YN = 'yersno';
+    const FIELD_TYPE_DATETIME = 'dateime';
+    const FIELD_TYPE_GAKEY = 'gakey';
+    const FIELD_TYPE_USE_CAPTCHA = 'use_captcha';
+
+    const GA_GLOBAL_KEY = '9999useGlobal9999';
+
+    public function __construct(
+        Permission $modelPermission,
+        Survey $modelSurvey,
+        LSYii_Application $yiiApp,
+        SessionData $sessionData,
+        PluginManager $yiiPluginManager,
+        LanguageConsistency $languageConsistency
+    ) {
+        $this->modelPermission = $modelPermission;
+        $this->modelSurvey = $modelSurvey;
+        $this->yiiApp = $yiiApp;
+        $this->sessionData = $sessionData;
+        $this->yiiPluginManager = $yiiPluginManager;
+        $this->languageConsistency = $languageConsistency;
+    }
+
+    /**
+     * Update
+     *
+     * @param int $surveyId
+     * @param array $input
+     * @throws ExceptionPersistError
+     * @throws ExceptionNotFound
+     * @throws ExceptionPermissionDenied
+     * @return array<array-key, mixed>
+     */
+    public function update($surveyId, $input)
+    {
+        $input = is_array($input) && !empty($input)
+            ? $input
+            : [];
+
+        $hasPermission = $this->modelPermission
+            ->hasSurveyPermission(
+                $surveyId,
+                'surveysettings',
+                'update'
+            );
+        if (!$hasPermission) {
+            throw new ExceptionPermissionDenied(
+                'Permission denied'
+            );
+        }
+
+        $survey = $this->modelSurvey->findByPk(
```

Comment:
If this is the only usage of $modelSurvey we might as well inject the $survey at call site instead?

Reply:
You are suggesting that the permission check and the loading of the survey should happen in the controller rather than in the service. In this case you could do that but I dont think its a good idea. We only then have to repeat this code each time we call this method. I dont see any gain from doing that. 

Having this logic in the service allows us to to define it once and test that it works correctly once. Both the permission check and the not found exception are already covered by unit tests.

Reply:
I was thinking in terms of functional core, imperative shell, and that it's easier to unit test service classes the fewer things you have to mock. So instead of doing

    $generalSettings->update($surveyId, $input)

you'd do

    $generalSettings->update(Survey::model()->findByPk($surveyId), $input);

You'd get a nullable type tho, and would still have to check for empty inside update(), probably throw a InvalidArgumentException or whatever.

This is what _composability_ (of functions) is all about. :)

Reply:
https://en.wikipedia.org/wiki/Composability

https://en.wikipedia.org/wiki/Function_composition_(computer_science)

Reply:
The permission model is used in two different places, so that one makes less sense to try to move out, I think.

Reply:
The gain is one less injection.

Reply:
One less injection in the constructor in return for code duplication in the controller and the REST command.

Reply:
Alright, I will resolve this then.